elastic/beats
# Add metrics_path as known hint for autodiscovery

This PR adds `metrics_path` as a known hint for autodiscovery.

This implements the suggestion of #13915 

### How to test
Run metricbeat as deamonset inside a k8s cluster having added the following config part for autodiscovery:
```
metricbeat.autodiscover:
  providers:
    - type: kubernetes
      host: ${NODE_NAME}
      hints.enabled: true
```
Note: run metricbeat with `-e -d autodiscover` to select only autodiscover related logs

Deploy a prometheus deployment:
```
kind: Deployment
metadata:
  name: prom-deployment
  namespace: kube-system
spec:
  selector:
    matchLabels:
      app: prometheus
  replicas: 1
  template:
    metadata:
      labels:
        app: prometheus
      annotations:
        co.elastic.metrics/module: prometheus
        co.elastic.metrics/hosts: '${data.host}:${data.port}'
        co.elastic.metrics/metrics_path: '/metrics/prometheus'
        co.elastic.metrics/period: 1m
    spec:
      containers:
      - name: prometheus
        image: prom/prometheus
        ports:
        - containerPort: 9090
```

Autodiscover should catch this event and include `metrics_path:/metrics/prometheus` in the configuration:
```
2019-10-10T08:16:07.148Z DEBUG [autodiscover] autodiscover/autodiscover.go:165 Got a start event: map[config:[0xc0008fcf60] host:172.17.0.34 id:e82994de-3de2-4355-afe2-0943d71ebd09.prometheus kubernetes:{"annotations":{"co":{"elastic":{"metrics/hosts":"${data.host}:${data.port}","metrics/metrics_path":"/metrics/prometheus","metrics/module":"prometheus","metrics/period":"1m"}}},"container":{"id":"be9be50a008e42f3f7e2d17edae2296ec0e5a009c3b3591ed711139f038246ea","image":"prom/prometheus","name":"prometheus","runtime":"docker"},"labels":{"app":"prometheus","pod-template-hash":"779f8d9569"},"namespace":"kube-system","node":{"name":"minikube"},"pod":{"name":"prom-deployment8-779f8d9569-6wjvx","uid":"e82994de-3de2-4355-afe2-0943d71ebd09"},"replicaset":{"name":"prom-deployment8-779f8d9569"}} meta:{"kubernetes":{"container":{"image":"prom/prometheus","name":"prometheus"},"labels":{"app":"prometheus","pod-template-hash":"779f8d9569"},"namespace":"kube-system","node":{"name":"minikube"},"pod":{"name":"prom-deployment8-779f8d9569-6wjvx","uid":"e82994de-3de2-4355-afe2-0943d71ebd09"},"replicaset":{"name":"prom-deployment8-779f8d9569"}}} port:9090 provider:958f1f51-570a-4e8d-9e13-f43503780543 start:true]
2019-10-10T08:16:07.148Z DEBUG [autodiscover] autodiscover/autodiscover.go:190 Generated config: map[enabled:true hosts:[172.17.0.34:9090] metrics_path:/metrics/prometheus metricsets:[collector] module:prometheus period:1m timeout:3s]
```
